contracts/governance/Staking/Staking.sol

```diff
@@ -90,32 +93,78 @@ contract Staking is WeightedStaking{
         require(balance > 0, "Staking:increaseStake: nothing staked yet until the given date. Use 'stake' instead.");
-        
+
+        //stake for the msg.sender if not specified otherwise
+        if(stakeFor == address(0)) {
+            stakeFor = msg.sender;
+        }
+
+        _increaseStake(amount, stakeFor, until);
+
+        _increaseDelegateStake(delegates[stakeFor][until], until, amount);
+    }
+
+    function _increaseStake(uint96 amount, address stakeFor, uint until) internal {
         //retrieve the SOV tokens
         bool success = SOVToken.transferFrom(msg.sender, address(this), amount);
         require(success);
-        
-        //stake for the msg.sender if not specified otherwise
-        if(stakeFor == address(0))
-            stakeFor = msg.sender;
-        
+
         //increase staked balance
+        uint96 balance = currentBalance(stakeFor, until);
         balance = add96(balance, amount, "Staking::increaseStake: balance overflow");
-        
+
         //update checkpoints
         _increaseDailyStake(until, amount);
-        _increaseDelegateStake(delegates[stakeFor][until], until, amount);
         _increaseUserStake(stakeFor, until, amount);
-        
+
         emit TokensStaked(stakeFor, amount, until, balance);
     }
-    
+
+    /**
+     * @notice stakes tokens according to the vesting schedule
+     * @param amount the amount of tokens to stake
+     * @param cliff the time interval to the first withdraw
+     * @param duration the staking duration
+     * @param intervalLength the length of each staking interval when cliff passed
+     * @param stakeFor the address to stake the tokens for or 0x0 if staking for oneself
+     * @param delegatee the address of the delegatee or 0x0 if there is none.
+     * */
+    function stakeTokens(
+        uint amount,
+        uint cliff,
+        uint duration,
+        uint intervalLength,
+        address stakeFor,
+        address delegatee
+    )
+        public
+    {
+        //stake them until lock dates according to the vesting schedule
+        //note: because staking is only possible in periods of 2 weeks, the total duration might
+        //end up a bit shorter than specified depending on the date of staking.
+        uint start = block.timestamp + cliff;
+        uint end = block.timestamp + duration;
+        uint numIntervals = (end - start) / intervalLength + 1;
+        uint stakedPerInterval = amount / numIntervals;
+        //stakedPerInterval might lose some dust on rounding. add it to the first staking date
+        if(numIntervals > 1) {
+            stake(uint96(amount - stakedPerInterval * (numIntervals-1)), start, stakeFor, delegatee);
+        }
+        //stake the rest in 4 week intervals
+        for(uint i = start + intervalLength; i <= end; i+= intervalLength) {
+            //stakes for itself, delegates to the owner
+            stake(uint96(stakedPerInterval), i, stakeFor, delegatee);
+        }
+    }
+
```

Comment:
has nothing to do with this PR, but I realized that we're not reducing the delegate stake when withdrawing tokens.

Reply:
fixed, tests updated